data/kernels/nlmeans.cl

```diff
@@ -70,16 +70,19 @@ nlmeans_dist(read_only image2d_t in, global float *U4, const int width, const in
   const int gidx = mad24(y, width, x);
   const float4 norm2 = (float4)(nL2, nC2, nC2, 1.0f);
 
-  // Reminder: q.x and q.y can be negative
   if(x >= width || y >= height) return;
-  if(x+q.x >= width || y+q.y >= height) return;
-  if(x+q.x < 0 || y+q.y < 0) return;
-
-  float4 p1 = read_imagef(in, sampleri, (int2)(x, y));
-  float4 p2 = read_imagef(in, sampleri, (int2)(x, y) + q);
-  float4 tmp = (p1 - p2)*(p1 - p2)*norm2;
-  float dist = tmp.x + tmp.y + tmp.z;
-  
+
+  // Reminder: q.x and q.y can be negative
+  float dist = 0.0f;
+  if(x+q.x < width && y+q.y < height
+    && x+q.x >= 0 && y+q.y >= 0)
+  {
+    float4 p1 = read_imagef(in, sampleri, (int2)(x, y));
```

Comment:
Thanks for this comment, I am very new at opencl, and I did not know that.
Isn't it an issue to do an out of bounds access if put the read outside any branch?
Otherwise I guess we can still use max and min function to make an access that is always in bounds..

As a general rule, does this mean we should try to make opencl code as constant time as possible? What amount of variation in time of execution can we tolerate? (For instance replacing all ifs of nlmeans_accu by computations without branches to be constant time is possible, but would probably make the code a bit harder to understand)

Reply:
Out of bounds access is another issue of OpenCL.

The problem with branch divergence is a bit different. All work group items need to process all branches. So in an if() statement the TRUE branch gets calculated as well as the FALSE branch in all work group items.

Reply:
Ok, thanks for the explanation :-)